### PR TITLE
maxtracks parameter for offline validation

### PR DESCRIPTION
**Pull request by hroskes**
So that the validation can stop after a specified number of tracks, rather than just a number of events.  Because the DMR plots count the number of modules per bin, rather than the number of tracks per bin, the number of tracks has to be the same when comparing two geometries.

### CODE REVIEW

**Comment by arunhep** [comment]
please test

File: Alignment/OfflineValidation/plugins/TrackerOfflineValidation.cc [hunk 1]
```diff
@@ -304,6 +304,9 @@ class TrackerOfflineValidation : public edm::EDAnalyzer {
   //sum hists are fit at the end using fitResiduals()
   std::vector<TH1*> toFit_;
 
+  unsigned long long ntracks_;
+  const unsigned long long maxtracks_;
```
Comment (ghellwig) [inline Alignment/OfflineValidation/plugins/TrackerOfflineValidation.cc]: @hroskes just a minor comment, but the convention in CMSSW would suggest this:
```
unsigned long long nTracks_;
const unsigned long long maxTracks_;
```

File: Alignment/OfflineValidation/plugins/TrackerOfflineValidation.cc [hunk 2]
```diff
@@ -386,6 +389,8 @@ TrackerOfflineValidation::TrackerOfflineValidation(const edm::ParameterSet& iCon
     useOverflowForRMS_(parSet_.getParameter<bool>("useOverflowForRMS")),
     dqmMode_(parSet_.getParameter<bool>("useInDqmMode")),
     moduleDirectory_(parSet_.getParameter<std::string>("moduleDirectoryInOutput")),
+    ntracks_(0),
+    maxtracks_(parSet_.getParameter<unsigned long long>("maxtracks")),
```
Comment (ghellwig) [inline Alignment/OfflineValidation/plugins/TrackerOfflineValidation.cc]: see above

**Review by ghellwig — CHANGES_REQUESTED** [review]
just a minor comment
otherwise  this PR is fine

**Comment by ghellwig** [comment]
please test

**Comment by ghellwig** [comment]
@hroskes one thing I realized: this implementation will not give the correct results in case of parallel jobs, right?
But it will instead give maxTracks * nJobs, if I am not mistaken.

**Comment by hroskes** [comment]
Oh you're right!  I did consider that but then I forgot when I came to implement it...  Ok I'll fix it now.

File: Alignment/OfflineValidation/python/TkAlAllInOneTool/genericValidation.py [hunk 3]
```diff
@@ -281,10 +281,13 @@ def __init__(self, valName, alignment, config):
...
         # if maxevents is not specified, cannot calculate number of events for
         # each parallel job, and therefore running only a single job
-        if int( self.general["maxevents"] ) == -1 and self.NJobs > 1:
+        if int( self.general["maxevents"] ) < 0 and self.NJobs > 1:
             msg = ("Maximum number of events (maxevents) not specified: "
                    "cannot use parallel jobs.")
             raise AllInOneError(msg)
+        if int( self.general["maxevents"] ) / self.NJobs != float( self.general["maxevents"] ) / self.NJobs:
+            msg = ("maxevents has to be divisible by parallelJobs")
+            raise AllInOneError(msg)
```
Comment (ghellwig) [inline Alignment/OfflineValidation/python/TkAlAllInOneTool/genericValidation.py]: @hroskes isn't that too restrictive as we could just take the remainder of tracks from the last job?

File: Alignment/OfflineValidation/python/TkAlAllInOneTool/offlineValidation.py [hunk 4]
```diff
@@ -45,6 +45,15 @@ def __init__(self, valName, alignment, config):
...
             raise AllInOneError(msg)
 
+        try:
+            self.NTracks = int(self.general["maxtracks"])
+            if self.NTracks < 0: raise ValueError
+        except ValueError:
+            raise AllInOneError("maxtracks has to be a positive integer, or 0 for no limit")
+
+        if self.NTracks / self.NJobs != float(self.NTracks) / self.NJobs:
+            raise AllInOneError("maxtracks has to be divisible by parallelJobs")
```
Comment (ghellwig) [inline Alignment/OfflineValidation/python/TkAlAllInOneTool/offlineValidation.py]: see above

File: Alignment/OfflineValidation/python/TkAlAllInOneTool/genericValidation.py (hunk 3 above)
Comment (hroskes) [reply Alignment/OfflineValidation/python/TkAlAllInOneTool/genericValidation.py]: I think it's easier this way, if you have 100k events and you want to have 3 jobs, you can just make it 99999 and you won't see the difference.  The previous behavior (for maxevents) was to just silently change it to 100000 / 3 = 33333 for each job.  At least this way you know exactly what's happening.

Comment (ghellwig) [reply Alignment/OfflineValidation/python/TkAlAllInOneTool/genericValidation.py]: ok

**Comment by ghellwig** [comment]
please test

**Comment by ghellwig** [comment]
+1
provided jenkins agrees

**Comment by davidlange6** [comment]
+1

**Pull request merged**